### PR TITLE
8263567: gtests don't terminate the VM safely

### PR DESCRIPTION
**Pull request by dholmes-ora**
Please review this fix to shutdown cleanly any JVMs created by gtests that aren't aborted.

Testing:
  - manual instrumentation to check all JVM construction and then deletion during gtest runs
  - local gtest testing
  - tiers 1-3 gtest testing

Thanks,
David

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263567](https://bugs.openjdk.java.net/browse/JDK-8263567): gtests don't terminate the VM safely

### Reviewers
 * [Thomas Stuefe](https://openjdk.java.net/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to f9a3bd8072ba1994cf1dd93229a87c5b3cf85c65
 * [Daniel D. Daugherty](https://openjdk.java.net/census#dcubed) (@dcubed-ojdk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4986/head:pull/4986` \
`$ git checkout pull/4986`

Update a local copy of the PR: \
`$ git checkout pull/4986` \
`$ git pull https://git.openjdk.java.net/jdk pull/4986/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4986`

View PR using the GUI difftool: \
`$ git pr show -t 4986`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4986.diff">https://git.openjdk.java.net/jdk/pull/4986.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/gtest/gtestMain.cpp
Comment (tstuefe) [inline test/hotspot/gtest/gtestMain.cpp:113]: underscore missing?

Comment (tstuefe) [reply test/hotspot/gtest/gtestMain.cpp:113]: and `_is_initialized` is now redundant and could be aliased to `_jvm != NULL`

File: test/hotspot/gtest/gtestMain.cpp
Comment (tstuefe) [inline test/hotspot/gtest/gtestMain.cpp:127]: init_jvm should just return the JavaVM pointer instead of a (rather boolean) int. The result int does not carry any information beyond "success". We can do the same with != NULL.

File: test/hotspot/gtest/gtestMain.cpp
Comment (tstuefe) [inline test/hotspot/gtest/gtestMain.cpp:220]: is the comma needed?

File: test/hotspot/gtest/unittest.hpp
Comment (tstuefe) [inline test/hotspot/gtest/unittest.hpp:99]: For better readibility I'd factor this out to an own function.

**Review by tstuefe — CHANGES_REQUESTED** [review]
Hi David,

looks good, small remarks inline and one general remark here:

I wonder whether we should not just always initialize the VM at launch and destroy it after finish, regardless of the tests we run. We even could abolish TEST and just always do TEST_VM.

While in theory TEST (without _VM) is a good idea because we save time on the VM initialization, in practice
1) usually one just runs always the full suite, or a subset of it which includes TEST_VM tests, so most of the time we pay for VM init anyway.
2) tests are done using copy+paste and use TEST_VM unnecessarily. OTOH I also already had TEST tests that needed VM stuff. I mean that maintaining the correct designation for the tests is cumbersome and error-prone.
3) The _VM flexibility, tied to order and selection of tests, means the time VM initialization happens is unpredictable, which makes analysis more difficult and behavior sometimes surprising.

We could simplify all this and the runner code by just unconditionally initializing the VM at start of the gtestlauncher. I think the benefits of differentiating between TEST and TEST_VM are slim in practice.

..Thomas

**Comment by dholmes-ora** [comment]
Hi Thomas,

On 4/08/2021 5:22 pm, Thomas Stuefe wrote:
> On Wed, 4 Aug 2021 05:55:11 GMT, David Holmes <dholmes@openjdk.org> wrote:
> 
>> Please review this fix to shutdown cleanly any JVMs created by gtests that aren't aborted.
>>
>> Testing:
>>    - manual instrumentation to check all JVM construction and then deletion during gtest runs
>>    - local gtest testing
>>    - tiers 1-3 gtest testing
>>
>> Thanks,
>> David
> 
> Hi David,
> 
> looks good, small remarks inline and one general remark here:
Thanks for taking a look at this. Sorry for the delayed follow-up. 
> I wonder whether we should not just always initialize the VM at launch and destroy it after finish, regardless of the tests we run. We even could abolish TEST and just always do TEST_VM.
> 
> While in theory TEST (without _VM) is a good idea because we save time on the VM initialization, in practice
> 1) usually one just runs always the full suite, or a subset of it which includes TEST_VM tests, so most of the time we pay for VM init anyway.
> 2) tests are done using copy+paste and use TEST_VM unnecessarily. OTOH I also already had TEST tests that needed VM stuff. I mean that maintaining the correct designation for the tests is cumbersome and error-prone.
> 3) The _VM flexibility, tied to order and selection of tests, means the time VM initialization happens is unpredictable, which makes analysis more difficult and behavior sometimes surprising.
> 
> We could simplify all this and the runner code by just unconditionally initializing the VM at start of the gtestlauncher. I think the benefits of differentiating between TEST and TEST_VM are slim in practice.

That may be true but I don't think it is for this issue/PR to try and address that. I can imagine that some basic TEST() cases assume a pristine environment with no chance of any external perturbation (such as by an initialized and thus operating VM). That said with the way tests get interleaved depending on exactly how you run them, it isn't clear that a TEST() may not run when the VM is already initialized anyway!

Thanks,
David

File: test/hotspot/gtest/gtestMain.cpp
Comment (dholmes-ora) [reply test/hotspot/gtest/gtestMain.cpp:113]: Yes good point - fixed.

File: test/hotspot/gtest/gtestMain.cpp
Comment (dholmes-ora) [reply test/hotspot/gtest/gtestMain.cpp:127]: I think the right thing to do is to report the JNI error that gets returned.

File: test/hotspot/gtest/gtestMain.cpp
Comment (dholmes-ora) [reply test/hotspot/gtest/gtestMain.cpp:220]: Which one? :) I wanted to convey that the fork and execution of `runUnitTestsInner` is done for each test. I will reword and restructure.

File: test/hotspot/gtest/unittest.hpp
Comment (dholmes-ora) [reply test/hotspot/gtest/unittest.hpp:99]: The problem is I don't have anywhere to put that function so that this macro can invoke it.

Comment (dholmes-ora) [reply test/hotspot/gtest/unittest.hpp:99]: To be more clear this is a .hpp file included by many .cpp files. If I put the function in the .hpp file and a specific .cpp file does not use the TEST_OTHER_VM macro then we have a function defined but not used and so hit `-Werror=unused-function`.
Is there some .cpp file I can put this in that would be globally accessible to all the test .cpp files?

File: test/hotspot/gtest/gtestMain.cpp
Comment (tstuefe) [reply test/hotspot/gtest/gtestMain.cpp]: The second one after "called" tickled my comma sense. But I am not a native speaker and may be wrong.

File: test/hotspot/gtest/gtestMain.cpp
Comment (tstuefe) [inline test/hotspot/gtest/gtestMain.cpp:121]: superfluous blank? (after first open bracket)

**Review by tstuefe — APPROVED** [review]
All good now. The fact that other_vm re-runs VM initialization tripped me off more than once btw :)

File: test/hotspot/gtest/gtestMain.cpp
Comment (dholmes-ora) [reply test/hotspot/gtest/gtestMain.cpp]: Fixed. Thanks

**Comment by dholmes-ora** [comment]
Could I get a second review please.

Thanks.

**Review by dcubed-ojdk — APPROVED** [review]
Thumbs up.

**Pull request closed without merging**
